Wrap DensityDist's random with generate_samples

lucianopaz opened the pull request:
This fixes issue #3553 by relying on `generate_samples` to make the random number generators aware of the distribution's shape. The old _incorrect_ behavior can still be achieved with some extra arguments during initialization.

lucianopaz committed:
Fix for 3553. Wrapped random with generate_samples. Added tests and extra control in DensityDist.

lucianopaz committed:
Add to RELEASE-NOTES.md

junpenglao commented on pymc3/distributions/distribution.py:
```diff
@@ -211,16 +211,101 @@ class DensityDist(Distribution):
+    ):
+        """
+        Parameters
+        ----------
```
NICE!

rpgoldman replied:
The one question I have -- and this is not a question about your change -- is whether we should have a default `shape` that is not `None` (i.e.,  that records the fact that the programmer did not specify shape).
I'm sure that the shape *is* most likely `()`, but worried about the case where it isn't, and the programmer/modeler has not realized that they need to specify the shape.
But this is a different issue, and should not delay merging this PR.

lucianopaz replied:
I would say no, we shouldn't add a special empty shape value. Ideally, the default random variable's shape should be deduced from its parameters. We are far from this goal. At the moment the shape information must be explicitly supplied to every distribution in pymc3, not just `DensityDist`. The default behavior package wise is to assume a scalar distribution.
When we start to move towards a more symbolic handling of shapes, we should address whether to reserve a special empty value or not.

lucianopaz commented:
@junpenglao, @twiecki  I have no idea what caused the test failure. Could be some deprecation that theano didn't account for?

lucianopaz committed:
Possible fix for glm utils test

junpenglao commented:
Seems to be a real test fail:

```python
self = <pymc3.tests.test_models_utils.TestUtils object at 0x7fd6abddbbe0>
    def test_dict_input(self):
        m, l = utils.any_to_tensor_and_labels(self.data.to_dict('dict'))
        self.assertMatrixLabels(m, l, mt=self.data[l].values, lt=l)
    
        m, l = utils.any_to_tensor_and_labels(self.data.to_dict('series'))
        self.assertMatrixLabels(m, l, mt=self.data[l].values, lt=l)
    
        m, l = utils.any_to_tensor_and_labels(self.data.to_dict('list'))
        self.assertMatrixLabels(m, l, mt=self.data[l].values, lt=l)
    
        inp = {k: tt.as_tensor_variable(v.values) for k, v in self.data.to_dict('series').items()}
>       m, l = utils.any_to_tensor_and_labels(inp)
pymc3/tests/test_models_utils.py:44: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pymc3/glm/utils.py:118: in any_to_tensor_and_labels
    x = tt.as_tensor_variable(x)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
x = array([[Subtensor{int64}.0, Subtensor{int64}.0],
       [Subtensor{int64}.0, Subtensor{int64}.0],
       [Subtensor{int64}.0, Subtensor{int64}.0]], dtype=object)
```

lucianopaz commented:
@junpenglao, but it's completely unrelated to the PR. In fact, I thought that maybe I should add a `.values` to fix the issue in 81e96b0, but the problem persisted.

junpenglao commented:
I am not clear about the context of this test, but looking at it naively I would say we dont need to convert `x` to tensor but instead just do `tt.concatenate`

rpgoldman reviewed (commented):
What about the notion of adding a `ShapeError` error class as I did in my PR?  It seems like this is enough of a thing that we might want it.

rpgoldman commented:
> I am not clear about the context of this test, but looking at it naively I would say we dont need to convert `x` to tensor but instead just do `tt.concatenate`

@lucianopaz is right: master is broken, too.

rpgoldman committed:
Extended the docstring.

rpgoldman reviewed (approved):
I added a little to the docstring, but other than that, looks great!

lucianopaz commented:
Thanks @rpgoldman. I'll edit the note you added though. The user supplied callable never sees the `dist_shape` parameter, it is only viewed by `generate_samples`.

lucianopaz committed:
Updated note

rpgoldman commented:
> Thanks @rpgoldman. I'll edit the note you added though. The user supplied callable never sees the `dist_shape` parameter, it is only viewed by `generate_samples`.

Maybe it would be helpful to give an example of this case? I'm not sure what the relationship of `generate_samples` to the `random` function is, and I'm not sure how the supplier of the `random` function to the `DensityDist` would be able to use the `dist_shape` if the generator can't see it.

I'm going to "activate" some more of the docstrings to help with this.

rpgoldman committed:
Ensure new docstrings appear in docs.

Make the docs for `generate_samples` and `draw_values` appear in the API manual pages.
Also add docs for other bits of distributions/distribution.py

lucianopaz commented:
Great @rpgoldman! I'll write up an example in the docstring

lucianopaz commented:
While writing an example I ran into an unexpected issue that this PR introduces. I'm setting the WIP flag now, as this is not ready to merge yet.

lucianopaz committed:
Added multidimensional dist tests that fail

lucianopaz committed:
Added fix and documented it

lucianopaz committed:
Updated release notes

lucianopaz commented:
Should be done now. There was a problem that popped out when I tried to write down examples using multidimensional distributions (shape problems never stop). I had to add in some more control flow but I hope this is now working well.

junpenglao commented on pymc3/distributions/distribution.py:
```diff
@@ -198,29 +198,315 @@ class DensityDist(Distribution):
         self.rand = random
+        self.wrap_random_with_dist_shape = wrap_random_with_dist_shape
+        self.check_shape_in_random = check_shape_in_random
+        self.pymc3_size_interpretation = pymc3_size_interpretation
```
Should we really expose these option to user? I think it might be better in terms of API to just always do these check and raise error when the input function does not have the correct shape handling.
Granted that would place the burden to user to provide a rigid function, but on the other hand there are much less option exposed.

WDYT? @twiecki

lucianopaz replied:
It is certainly not nice to have these extra parameters here. They only allow the users to go back to the old behavior. I don't know if this is a values enough reason to include them into the construction signature

rpgoldman replied:
@junpenglao I think if we were to remove these we should just go go always using `wrap_random_with_dist_shape=True` (or the equivalent).  This might be backwards-incompatible, but the backwards way was wrong, as far as I can tell.

junpenglao replied:
I dont think we need to be backward compatible in this case if the old way was wrong/incomplete.

lucianopaz replied:
I removed the `pymc3_size_interpretation` kwarg

rpgoldman reviewed (commented):
Looks good to me!

The two meanings of the same keyword argument (`size`) is really yucky, but I don't have an idea how to fix it.

lucianopaz committed:
Remove pymc3_size_interpretation kwarg

lucianopaz committed:
Added point parameter to rand call

rpgoldman commented:
@lucianopaz Should I rebase this, update RELEASE-NOTES and merge? Or would you like to?

@junpenglao and I both gave it the 👍🏻, and it still looks good after the updates (and maybe the rebase means the first 2 of the additional commits are not necessary?).

lucianopaz commented:
Thanks @rpgoldman. I'd be able to do this on Monday so feel free to go ahead with the rebase and merge.

rpgoldman commented:
Rebased.  Now waiting to make sure the Travis tests still pass. If so, will merge.

rpgoldman commented:
Thanks, @lucianopaz !
